Deflake tests in `staging/src/k8s.io/kube-aggregator/pkg/apiserver`

gjkim42 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind failing-test
/kind flake

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This PR addresses two issues in `handler_discovery_test.go`.
1. Updating and reading the bool-type variable is not thread-safe. We can simply fix this one.
2. (This is quite tricky to address) It is hard to know that `discoveryManager` completes the sync on all items in the queue.

More details on issue 2 below:
`waitForEmptyQueue` cannot guarantee that `discoveryManager` completes the sync on all items in the queue. It just guarantee that all items have been started to sync.

This PR adds `waitForQueueComplete` and implements `completerWorkqueue` to check if the workqueue is complete.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #115858 

```sh
$ stress ./apiserver.test                    
5s: 0 runs so far, 0 failures       
10s: 12 runs so far, 0 failures     
15s: 24 runs so far, 0 failures     
...
3m50s: 445 runs so far, 0 failures
3m55s: 456 runs so far, 0 failures
4m0s: 461 runs so far, 0 failures
```

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
none
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

aojea commented:
in a super quick look, I can observe that you have getters to the controller cache, maybe that can give you a way to confirm that all items were processed ?

alexzielenski commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go:
```diff
@@ -151,17 +152,18 @@ func TestDirty(t *testing.T) {
 
 	go aggregatedManager.Run(testCtx.Done())
 	require.True(t, waitForEmptyQueue(testCtx.Done(), aggregatedManager))
+	aggregatedManager.dirtyAPIServiceQueue.ShutDownWithDrain()
```
While this blocks until all items in the queue have synced, I don't think it is the right general solution. `waitForEmptyQueue` needs to be modified to wait for queue to stop processing, otherwise the other tests that use `waitForEmptyQueue` will still be affected

gjkim42 replied:
I think so too. I am trying to solve this with more general way. See https://github.com/kubernetes/kubernetes/pull/115859#issuecomment-1434881423

gjkim42 commented:
@aojea

> in a super quick look, I can observe that you have getters to the controller cache, maybe that can give you a way to confirm that all items were processed ?

I think the way (using cache) looks not clear for now.

Actually, I am struggling to solve this issue with more general way. (the current fix looks just a hack for me)

I feel that there are a few workarounds,
1. waiting for the desired state with timeout in tests
2. adding processingCount to the controller
3. adding IsEmpty(Len is 0 and is not processing anything) to queue.Interface

3 looks better approach for me... what do you think?

gjkim42 commented:
> 3\. adding IsEmpty(Len is 0 and is not processing anything) method to queue.Interface

applied 3.

One concern is that I am not sure it is ok to add a method to `queue.Interface`

gjkim42 commented:
> > 3. adding IsEmpty(Len is 0 and is not processing anything) method to queue.Interface
> 
> applied 3.
> 
> One concern is that I am not sure it is ok to add a method to `queue.Interface`

There are some other implementations for `queue.Interface`...

gjkim42 commented:
added completerWorkQueue to check if the workqueue is complete for testing purpose.

gjkim42 committed:
Fix a data race in TestDirty

This uses atomic.Bool as updating and reading a boolean-type variable
concurrently is not thread-safe.

gjkim42 committed:
Deflake tests in staging/src/k8s.io/kube-aggregator/pkg/apiserver

`waitForEmptyQueue` cannot guarantee that all items in the queue have
been synced completely but guarantee that all items have been started.

This adds `waitForQueueComplete` and implements `completerWorkqueue` to
check if the workqueue is complete to deflake the tests in
staging/src/k8s.io/kube-aggregator/pkg/apiserver.

gjkim42 commented:
/priority important-longterm

@alexzielenski 

Could you PTAL?

alexzielenski commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go:
```diff
@@ -437,3 +442,48 @@ func fetchPath(handler http.Handler, etag string) (*http.Response, []byte, *apid
+	q.RateLimitingInterface.Add(item)
+}
+
+func (q *completerWorkqueue) AddAfter(item interface{}, duration time.Duration) {
+	q.Add(item)
+}
+
+func (q *completerWorkqueue) AddRateLimited(item interface{}) {
+	q.Add(item)
+}
```
This feels weird to me, but I know neither of these ever are called by discovery manager, since [`syncAPIService`](https://github.com/kubernetes/kubernetes/blob/40c5b1b7c37804f5c8656dffadf7f086daa0a9ba/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go#L344) unconditionally returns nil. (Discovery Manager refreshes all services on minute interval unconditionally, so does not make use of rate limiter)

I'm wondering if this or a future PR should refactor it to not use the `workqueue.RateLimitingInterface` and instead use normal `workqueue.Interface`. I think we can keep that in mind for a separate PR, to retain this PR's nice property of only changing test files.

alexzielenski commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go:
```diff
@@ -361,18 +366,18 @@ func TestNotModified(t *testing.T) {
-	require.True(t, waitForEmptyQueue(testCtx.Done(), aggregatedManager))
+	require.True(t, waitForQueueComplete(testCtx.Done(), aggregatedManager))
 }
 
```
I created this test which deterministically caught the flake before your change, and is now working after your change. It makes the sync handler take a long time to catch the queue being empty, but handler still syncing. I think maybe it would be good to add:

```go
// Shows that waitForQueueComplete also waits for syncing to
// complete by artificially making the sync handler take a long time
func TestWaitForSync(t *testing.T) {
	pinged := atomic.Bool{}
	service := discoveryendpoint.NewResourceManager()
	aggregatedResourceManager := discoveryendpoint.NewResourceManager()

	aggregatedManager := newDiscoveryManager(aggregatedResourceManager)

	aggregatedManager.AddAPIService(&apiregistrationv1.APIService{
		ObjectMeta: metav1.ObjectMeta{
			Name: "v1.stable.example.com",
		},
		Spec: apiregistrationv1.APIServiceSpec{
			Group:   "stable.example.com",
			Version: "v1",
			Service: &apiregistrationv1.ServiceReference{
				Name: "test-service",
			},
		},
	}, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		time.Sleep(3 * time.Second)
		pinged.Store(true)
		service.ServeHTTP(w, r)
	}))
	testCtx, cancel := context.WithCancel(context.Background())
	defer cancel()

	go aggregatedManager.Run(testCtx.Done())
	require.True(t, waitForQueueComplete(testCtx.Done(), aggregatedManager))

	// immediately check for ping, since Run() should block for local services
	if !pinged.Load() {
		t.Errorf("service handler never pinged")
	}
}
```

gjkim42 replied:
added it

alexzielenski reviewed (approved):
Thanks @gjkim42! I really like this solution with the test-specific completerWorkqueue. I think it solves this problem fairly cleanly. 

One minor comment about the RateLimitingInterface that may not need to be addressed here.

/lgtm
/assign @liggitt 
/cc @liggitt 
I think this test-only change will also need to be back ported to 1.26. My change to patch a race #115302 introduced this flake.

aojea commented:
> I think the way (using cache) looks not clear for now.

then forget my comment, I most probably misunderstood it, seems Alex is taking on the review , he understand this better than me

gjkim42 commented:
updated the commit message and rebased the PR.

gjkim42 commented:
/priority important-soon

`pull-kubernetes-unit` can be flaky.

gjkim42 commented:
/remove-priority important-longterm

cici37 commented:
/triage accepted

gjkim42 commented:
> @alexzielenski can you take a final look at the most recent updates?

added a test that you suggested and changed the commit message.
Could you PTAL?

alexzielenski commented:
/lgtm

gjkim42 commented:
@liggitt 
Could you take a look?

> I think this test-only change will also need to be back ported to 1.26. My change to patch a race #115302 introduced this flake.

And we think that this needs to be backported.

aojea commented:
> in a super quick look, I can observe that you have getters to the controller cache, maybe that can give you a way to confirm that all items were processed ?

```diff
diff --git a/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go b/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go
index 9bd48972404..c07993779d7 100644
--- a/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go
+++ b/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go
@@ -18,6 +18,7 @@ package apiserver
 
 import (
        "context"
+       "fmt"
        "net/http"
        "net/http/httptest"
        "strconv"
@@ -44,12 +45,14 @@ func newDiscoveryManager(rm discoveryendpoint.ResourceManager) *discoveryManager
        return NewDiscoveryManager(rm).(*discoveryManager)
 }
 
-// Returns true if the queue of services to sync empty this means everything has
-// been reconciled and placed into merged document
-func waitForEmptyQueue(stopCh <-chan struct{}, dm *discoveryManager) bool {
+// Returns true when it reachs the number of the cached services
+func waitForCacheNumber(stopCh <-chan struct{}, n int, dm *discoveryManager) bool {
        return cache.WaitForCacheSync(stopCh, func() bool {
                // Once items have successfully synced they are removed from queue.
-               return dm.dirtyAPIServiceQueue.Len() == 0
+               dm.resultsLock.Lock()
+               defer dm.resultsLock.Unlock()
+               fmt.Println("DEBUG", len(dm.cachedResults), n)
+               return len(dm.cachedResults) == n
        })
 }
 
@@ -63,7 +66,6 @@ func TestBasic(t *testing.T) {
        service2.SetGroups(apiGroup2.Items)
        aggregatedResourceManager := discoveryendpoint.NewResourceManager()
        aggregatedManager := newDiscoveryManager(aggregatedResourceManager)
-
        for _, g := range apiGroup1.Items {
                for _, v := range g.Versions {
                        aggregatedManager.AddAPIService(&apiregistrationv1.APIService{
@@ -103,7 +105,7 @@ func TestBasic(t *testing.T) {
 
        go aggregatedManager.Run(testCtx.Done())
 
-       require.True(t, waitForEmptyQueue(testCtx.Done(), aggregatedManager))
+       require.True(t, waitForCacheNumber(testCtx.Done(), 2, aggregatedManager))
 
        response, _, parsed := fetchPath(aggregatedResourceManager, "")
        if response.StatusCode != 200 {
@@ -159,7 +161,9 @@ func TestDirty(t *testing.T) {
        defer cancel()
 
        go aggregatedManager.Run(testCtx.Done())
-       require.True(t, waitForEmptyQueue(testCtx.Done(), aggregatedManager))
+       require.True(t, waitForCacheNumber(testCtx.Done(), 1, aggregatedManager))
+       // time.Sleep(1 * time.Second)
+       fmt.Println("POST", aggregatedManager.apiServices, aggregatedManager.cachedResults)
 
        // immediately check for ping, since Run() should block for local services
        if !pinged.Load() {
@@ -211,7 +215,7 @@ func TestRemoveAPIService(t *testing.T) {
                aggregatedManager.RemoveAPIService(s.Name)
        }
 
-       require.True(t, waitForEmptyQueue(testCtx.Done(), aggregatedManager))
+       require.True(t, waitForCacheNumber(testCtx.Done(), 0, aggregatedManager))
 
        response, _, parsed := fetchPath(aggyService, "")
        if response.StatusCode != 200 {
@@ -293,7 +297,7 @@ func TestLegacyFallback(t *testing.T) {
        defer cancel()
 
        go aggregatedManager.Run(testCtx.Done())
-       require.True(t, waitForEmptyQueue(testCtx.Done(), aggregatedManager))
+       require.True(t, waitForCacheNumber(testCtx.Done(), 1, aggregatedManager))
 
        // At this point external services have synced. Check if discovery document
        // includes the legacy resources
@@ -362,7 +366,7 @@ func TestNotModified(t *testing.T) {
 
        // Important to wait here to ensure we prime the cache with the initial list
        // of documents in order to exercise 304 Not Modified
-       require.True(t, waitForEmptyQueue(testCtx.Done(), aggregatedManager))
+       require.True(t, waitForCacheNumber(testCtx.Done(), 1, aggregatedManager))
 
        // Now add all groups. We excluded one group before so that AllServicesSynced
        // could include it in this round. Now, if AllServicesSynced ever returns
@@ -373,7 +377,7 @@ func TestNotModified(t *testing.T) {
        }
 
        // This would wait the full timeout on 1.26.0.
-       require.True(t, waitForEmptyQueue(testCtx.Done(), aggregatedManager))
+       require.True(t, waitForCacheNumber(testCtx.Done(), 1, aggregatedManager))
 }
 ```

I tried the cache approach and looks easier than the queue

```
stress ./apiserver.test -test.run TestDirty
5s: 999 runs so far, 0 failures
10s: 2071 runs so far, 0 failures
15s: 3137 runs so far, 0 failures
20s: 4173 runs so far, 0 failures
25s: 5219 runs so far, 0 failures
30s: 6275 runs so far, 0 failures
35s: 7326 runs so far, 0 failures
40s: 8383 runs so far, 0 failures
45s: 9470 runs so far, 0 failures
50s: 10549 runs so far, 0 failures
```

alexzielenski commented:
@aojea while that is a good approach I think if we in the future added a test for updating a pre-existing entry (which I'm likely to add very soon) this might not be sufficient if I'm reading correctly?

liggitt commented:
/approve

liggitt commented:
looks like there are verify errors to resolve

/hold

gjkim42 commented:
> looks like there are verify errors to resolve

That is because https://github.com/kubernetes/kubernetes/pull/115770 has been merged which introduces a new test to fix.

Now, I rebased this PR and fixed the newly created test as well.

```
$ stress ./apiserver.test
5s: 0 runs so far, 0 failures       
10s: 12 runs so far, 0 failures     
15s: 12 runs so far, 0 failures
...
1h0m0s: 4582 runs so far, 0 failures
```

@alexzielenski @liggitt 

Could you re-LGTM on this?

liggitt commented:
/lgtm

liggitt commented:
/hold cancel

aojea commented:
> @aojea while that is a good approach I think if we in the future added a test for updating a pre-existing entry (which I'm now likely to add very soon) your alternative not be sufficient if I'm reading correctly?

@alexzielenski it was just an alternative because I saw your comment[ in the workqueue complete](https://github.com/kubernetes/kubernetes/pull/115859#discussion_r1110177336), but I don't have strong opinion, the current option LGTM too
